api/v1beta2/ratelimitpolicy_types.go

```diff
@@ -0,0 +1,217 @@
+	// https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io/v1beta1.HTTPRouteSpec
+	// +optional
+	Matches []gatewayapiv1alpha2.HTTPRouteMatch `json:"matches,omitempty"`
+}
+
+// Limit represents a complete rate limit configuration
+type Limit struct {
+	// RouteSelectors defines semantics for matching an HTTP request based on conditions
+	// +optional
+	RouteSelectors []RouteSelector `json:"routeSelectors,omitempty"`
+
+	// When holds the list of conditions for the policy to be enforced.
+	// Called also "soft" conditions as route selectors must also match
+	// +optional
+	When []WhenCondition `json:"when,omitempty"`
+
+	// Counters defines additional rate limit counters based on context qualifiers and well known selectors
+	// TODO Document properly "Well-known selector" https://github.com/Kuadrant/architecture/blob/main/rfcs/0001-rlp-v2.md#well-known-selectors
+	// +optional
+	Counters []ContextSelector `json:"counters,omitempty"`
+
+	// Rates holds the list of limit rates
+	// +optional
+	Rates []Rate `json:"rates,omitempty"`
+}
+
+// RateLimitPolicySpec defines the desired state of RateLimitPolicy
+type RateLimitPolicySpec struct {
+	// TargetRef identifies an API object to apply policy to.
+	TargetRef gatewayapiv1alpha2.PolicyTargetReference `json:"targetRef"`
+
+	// Limits holds the struct of limits indexed by a unique name
+	// +optional
+	Limits map[string]Limit `json:"limits,omitempty"`
+}
+
+// RateLimitPolicyStatus defines the observed state of RateLimitPolicy
+type RateLimitPolicyStatus struct {
+	// ObservedGeneration reflects the generation of the most recently observed spec.
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+	// Represents the observations of a foo's current state.
+	// Known .status.conditions.type are: "Available"
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=type
+	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type" protobuf:"bytes,1,rep,name=conditions"`
+}
+
+func (s *RateLimitPolicyStatus) Equals(other *RateLimitPolicyStatus, logger logr.Logger) bool {
+	if s.ObservedGeneration != other.ObservedGeneration {
+		diff := cmp.Diff(s.ObservedGeneration, other.ObservedGeneration)
+		logger.V(1).Info("ObservedGeneration not equal", "difference", diff)
+		return false
+	}
+
+	// Marshalling sorts by condition type
+	currentMarshaledJSON, _ := common.ConditionMarshal(s.Conditions)
+	otherMarshaledJSON, _ := common.ConditionMarshal(other.Conditions)
+	if string(currentMarshaledJSON) != string(otherMarshaledJSON) {
+		diff := cmp.Diff(string(currentMarshaledJSON), string(otherMarshaledJSON))
+		logger.V(1).Info("Conditions not equal", "difference", diff)
+		return false
+	}
+
+	// TODO(eastizle): reflect.DeepEqual does not work well with lists without order
+	if !reflect.DeepEqual(s.GatewaysRateLimits, other.GatewaysRateLimits) {
```

Comment:
I wonder if order matters. If it does matter, do we want to enforce same order? I'm guessing "yes, we do", right?

Reply:
I have removed the `GatewaysRateLimits` struct from the status section of the CRD. I guess this is still a valid feature, but needs to be refactored entirely. We can add items to the status later when we implement the feature. Does it sound good to you?